File: mobile/scripts/auto_translate.js
const fs = require('fs');
const path = require('path');

const LIB_DIR = '/Users/os2026/Downloads/Reservatior/mobile/lib';
const LOCALES_DIR = '/Users/os2026/Downloads/Reservatior/client/src/locales';
const MOBILE_LOCALES_DIR = '/Users/os2026/Downloads/Reservatior/mobile/assets/translations';

const textPattern1 = /Text\('([^$'\\]+)'\)/g;
const textPattern2 = /Text\("([^$"\\]+)"\)/g;
const constTextPattern1 = /const\s+Text\('([^$'\\]+)'\)/g;
const constTextPattern2 = /const\s+Text\("([^$"\\]+)"\)/g;
const labelPattern1 = /(labelText|hintText|tooltip):\s*'([^$'\\]+)'/g;
const labelPattern2 = /(labelText|hintText|tooltip):\s*"([^$"\\]+)"/g;

function toKey(text) {
    let safe = text.replace(/[^a-zA-Z0-9]/g, '_').toLowerCase();
    safe = safe.replace(/_+/g, '_').replace(/^_|_$/g, '');
    if (!safe) return null;
    return "mobile.auto." + safe.substring(0, 40);
}

function processDartFile(filepath, newKeys) {
    let content = fs.readFileSync(filepath, 'utf8');
    const original = content;

    const hasImport = content.includes("import 'package:easy_localization/easy_localization.dart';");

    function repl(match, p1) {
        if (!p1.trim()) return match;
        const key = toKey(p1);
        if (!key) return match;
        newKeys[key] = p1;
        return `Text('${key}'.tr())`;
    }

    function replLabel(match, attr, val) {
        if (!val.trim()) return match;
        const key = toKey(val);
        if (!key) return match;
        newKeys[key] = val;
        return `${attr}: '${key}'.tr()`;
    }

    content = content.replace(constTextPattern1, repl);
    content = content.replace(constTextPattern2, repl);
    content = content.replace(textPattern1, repl);
    content = content.replace(textPattern2, repl);
    content = content.replace(labelPattern1, replLabel);
    content = content.replace(labelPattern2, replLabel);

    if (content !== original) {
        if (!hasImport) {
            const imports = content.match(/^import\s+.*?;/gm);
            if (imports && imports.length > 0) {
                const lastImport = imports[imports.length - 1];
                content = content.replace(lastImport, lastImport + "\nimport 'package:easy_localization/easy_localization.dart';");
            } else {
                content = "import 'package:easy_localization/easy_localization.dart';\n" + content;
            }
        }

        content = content.replace(/\bconst\s+(Row|Column|Padding|Center|Expanded|Align|SizedBox|ListTile|TextFormField|Card|Container|Icon|ElevatedButton|OutlinedButton|TextButton|FloatingActionButton|Drawer|Scaffold|AppBar|BottomNavigationBar|TabBar|Tab|ListView|GridView|SingleChildScrollView|Stack|Positioned|Flexible|FractionallySizedBox|Spacer|Divider|CircularProgressIndicator|LinearProgressIndicator|Image|Tooltip|Text|InputDecoration|EdgeInsets|BoxDecoration|BorderRadius|Border|BoxShadow|Color|Colors)\b/g, '$1');
        content = content.replace(/\bconst\s+\[/g, '[');
        content = content.replace(/children:\s*const\s*\[/g, 'children: [');

        fs.writeFileSync(filepath, content, 'utf8');
        return true;
    }
    return false;
}

function walk(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filepath = path.join(dir, file);
        if (fs.statSync(filepath).isDirectory()) {
            walk(filepath, fileList);
        } else if (file.endsWith('.dart')) {
            fileList.push(filepath);
        }
    }
    return fileList;
}

function syncKeys(dir, newKeys) {
    if (!fs.existsSync(dir)) return;
    const files = fs.readdirSync(dir);
    for (const file of files) {
        if (file.endsWith('.json')) {
            const fp = path.join(dir, file);
            let data;
            try {
                data = JSON.parse(fs.readFileSync(fp, 'utf8'));
            } catch (e) {
                console.log(`Could not read ${fp}`);
                continue;
            }
            let added = 0;
            for (const [k, v] of Object.entries(newKeys)) {
                if (!(k in data)) {
                    data[k] = v;
                    added++;
                }
            }
            if (added > 0) {
                fs.writeFileSync(fp, JSON.stringify(data, null, 2) + '\n', 'utf8');
                console.log(`Added ${added} keys to ${fp}`);
            }
        }
    }
}

function main() {
    const newKeys = {};
    let changedFiles = 0;

    const files = walk(LIB_DIR);
    for (const file of files) {
        if (processDartFile(file, newKeys)) {
            changedFiles++;
        }
    }

    console.log(`Changed ${changedFiles} Dart files.`);
    const keyCount = Object.keys(newKeys).length;
    console.log(`Found ${keyCount} new translation keys.`);

    if (keyCount > 0) {
        syncKeys(LOCALES_DIR, newKeys);
        syncKeys(MOBILE_LOCALES_DIR, newKeys);
    }
}

main();

File: mobile/scripts/auto_translate.py
import os
import re
import json

LIB_DIR = '/Users/os2026/Downloads/Reservatior/mobile/lib'
LOCALES_DIR = '/Users/os2026/Downloads/Reservatior/client/src/locales'
MOBILE_LOCALES_DIR = '/Users/os2026/Downloads/Reservatior/mobile/assets/translations'

# Regex to find Text('...') or Text("...") without interpolation
text_pattern1 = re.compile(r"Text\('([^$'\\]+)'\)")
text_pattern2 = re.compile(r'Text\("([^$"\\]+)"\)')
const_text_pattern1 = re.compile(r"const\s+Text\('([^$'\\]+)'\)")
const_text_pattern2 = re.compile(r'const\s+Text\("([^$"\\]+)"\)')

# Regex to find labelText: '...' etc
label_pattern1 = re.compile(r"(labelText|hintText|tooltip):\s*'([^$'\\]+)'")
label_pattern2 = re.compile(r'(labelText|hintText|tooltip):\s*"([^$"\\]+)"')

def to_key(text):
    safe = "".join([c.lower() if c.isalnum() else '_' for c in text])
    safe = re.sub(r'_+', '_', safe).strip('_')
    if not safe:
        return None
    return "mobile.auto." + safe[:40]

def process_dart_file(filepath, new_keys):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original = content
    
    # Check if easy_localization is imported
    has_import = "import 'package:easy_localization/easy_localization.dart';" in content
    
    def repl(m):
        val = m.group(1)
        if len(val.strip()) == 0: return m.group(0)
        key = to_key(val)
        if not key: return m.group(0)
        new_keys[key] = val
        return f"Text('{key}'.tr())"
        
    def repl_label(m):
        attr = m.group(1)
        val = m.group(2)
        if len(val.strip()) == 0: return m.group(0)
        key = to_key(val)
        if not key: return m.group(0)
        new_keys[key] = val
        return f"{attr}: '{key}'.tr()"

    content = const_text_pattern1.sub(repl, content)
    content = const_text_pattern2.sub(repl, content)
    content = text_pattern1.sub(repl, content)
    content = text_pattern2.sub(repl, content)
    content = label_pattern1.sub(repl_label, content)
    content = label_pattern2.sub(repl_label, content)

    if content != original:
        if not has_import:
            # Add import after the last import
            imports = re.findall(r"^import\s+.*?;", content, re.MULTILINE)
            if imports:
                last_import = imports[-1]
                content = content.replace(last_import, last_import + "\nimport 'package:easy_localization/easy_localization.dart';")
            else:
                content = "import 'package:easy_localization/easy_localization.dart';\n" + content
                
        # Aggressive const removal to avoid build errors:
        # 1. Remove "const " before any widget name
        content = re.sub(r'\bconst\s+(Row|Column|Padding|Center|Expanded|Align|SizedBox|ListTile|TextFormField|Card|Container|Icon|ElevatedButton|OutlinedButton|TextButton|FloatingActionButton|Drawer|Scaffold|AppBar|BottomNavigationBar|TabBar|Tab|ListView|GridView|SingleChildScrollView|Stack|Positioned|Flexible|FractionallySizedBox|Spacer|Divider|CircularProgressIndicator|LinearProgressIndicator|Image|Tooltip|Text|InputDecoration|EdgeInsets|BoxDecoration|BorderRadius|Border|BoxShadow|Color|Colors)\b', r'\1', content)
        # 2. Remove "const [" to "["
        content = re.sub(r'\bconst\s+\[', r'[', content)
        # 3. Remove "children: const ["
        content = re.sub(r'children:\s*const\s*\[', r'children: [', content)

        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return True
    return False

def sync_keys(directory, new_keys):
    if not os.path.exists(directory): return
    for filename in os.listdir(directory):
        if filename.endswith('.json'):
            fp = os.path.join(directory, filename)
            try:
                with open(fp, 'r', encoding='utf-8') as f:
                    data = json.load(f)
            except Exception as e:
                print(f"Could not read {fp}, skipping.")
                continue
                
            added = 0
            for k, v in new_keys.items():
                if k not in data:
                    data[k] = v
                    added += 1
                    
            if added > 0:
                with open(fp, 'w', encoding='utf-8') as f:
                    json.dump(data, f, indent=2, ensure_ascii=False)
                    f.write('\n')
                print(f"Added {added} keys to {fp}.")

def main():
    new_keys = {}
    changed_files = 0
    
    for root, _, files in os.walk(LIB_DIR):
        for file in files:
            if file.endswith('.dart'):
                fp = os.path.join(root, file)
                if process_dart_file(fp, new_keys):
                    changed_files += 1
                    
    print(f"Changed {changed_files} Dart files.")
    print(f"Found {len(new_keys)} new translation keys.")
    
    if len(new_keys) > 0:
        sync_keys(LOCALES_DIR, new_keys)
        sync_keys(MOBILE_LOCALES_DIR, new_keys)

if __name__ == '__main__':
    main()
